### PR TITLE
Fix Egress IP Deletion Handler to Prevent OVN Policy Leaks

### PR DESCRIPTION
**Pull request by flavio-fernandes**
Resolved an issue where the Egress IP deletion handler misidentified affected pods, leading to unintended leaks of OVN logical router policies and SNATs. This occurred in scenarios with matchExpressions that only excluded keys and values.

Reported-at: https://issues.redhat.com/browse/OCPBUGS-24055

### CODE REVIEW

File: go-controller/pkg/ovn/egressip_test.go [hunk 1]
```diff
@@ -5094,8 +5094,15 @@ var _ = ginkgo.Describe("OVN master EgressIP Operations", func() {
...
 					},
+					// PodSelector: metav1.LabelSelector{
+					// 	MatchLabels: egressPodLabel,
+					// },
+					// https://issues.redhat.com/browse/OCPBUGS-24055
 					PodSelector: metav1.LabelSelector{
-						MatchLabels: egressPodLabel,
+						MatchExpressions: []metav1.LabelSelectorRequirement{{
+							Key:      "someNonExistingKey",
+							Operator: metav1.LabelSelectorOpDoesNotExist,
```
Comment (tssurya) [inline go-controller/pkg/ovn/egressip_test.go:5104]: let's also do a test for `Key` `NotIn` `Values being a string of stuff`

File: go-controller/pkg/ovn/egressip.go [hunk 2]
```diff
@@ -389,8 +389,8 @@ func (oc *DefaultNetworkController) reconcileEgressIPPod(old, new *v1.Pod) (err
 				return err
 			}
 			if !podSelector.Empty() {
-				newMatches := podSelector.Matches(newPodLabels)
-				oldMatches := podSelector.Matches(oldPodLabels)
+				newMatches := new != nil && podSelector.Matches(newPodLabels)
```
Comment (tssurya) [inline go-controller/pkg/ovn/egressip.go:392]: tiny bit confusing that we are not using oldPod and newPod which is what we use to obtain oldPodLabels and newPodLabes, but makes sense here that we don't want to load matches like "in" && "notin" and anything really if the pod does not exist in the first place either old or new.

File: go-controller/pkg/ovn/egressip_test.go [hunk 3]
```diff
@@ -5094,8 +5094,15 @@ var _ = ginkgo.Describe("OVN master EgressIP Operations", func() {
 					EgressIPs: []string{
 						updatedEgressIP.String(),
 					},
+					// PodSelector: metav1.LabelSelector{
+					// 	MatchLabels: egressPodLabel,
+					// },
+					// https://issues.redhat.com/browse/OCPBUGS-24055
```
Comment (tssurya) [inline go-controller/pkg/ovn/egressip_test.go:5100]: nit: I would refrain from adding JIRA bug links to upstream code (jira is very redhat specific) ; unless there is some TODO in place you want to get back to.
Also why are we commenting this code?

**Review by tssurya — COMMENTED** [review]
overall lgtm
are there other spots where we do such logic or are we good in all places?

File: go-controller/pkg/ovn/egressip_test.go (hunk 3 above)
Comment (kyrtapz) [reply go-controller/pkg/ovn/egressip_test.go:5100]: +1
Additionally that jira card is not public

Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip_test.go:5100]: > +1 Additionally that jira card is not public

TY both. I will take care of this soon.

**Comment by flavio-fernandes** [comment]
> overall lgtm are there other spots where we do such logic or are we good in all places?

Yes, @tssurya . Thank you for asking.  I looked at these reconciling functions and think they are all okay.

```
❯ grep -inr 'reconcile' --include "*" 2>/dev/null | grep -i "func " | grep -v vendor/ | grep -i "new "

pkg/clustermanager/pod/allocator.go:97:func (a *PodAllocator) Reconcile(old, new *corev1.Pod) error {
pkg/clustermanager/pod/allocator.go:124:func (a *PodAllocator) reconcile(old, new *corev1.Pod, releaseFromAllocator bool) error {
pkg/clustermanager/pod/allocator.go:164:func (a *PodAllocator) reconcileForNAD(old, new *corev1.Pod, nad string, network *nettypes.NetworkSelectionElement, releaseIPsFromAllocator bool) error {
pkg/clustermanager/egressip_controller.go:892:func (eIPC *egressIPClusterController) reconcileEgressIP(old, new *egressipv1.EgressIP) (err error) {
pkg/clustermanager/egressip_controller.go:1474:func (eIPC *egressIPClusterController) reconcileCloudPrivateIPConfig(old, new *ocpcloudnetworkapi.CloudPrivateIPConfig) error {
pkg/ovn/egressip.go:74:func (oc *DefaultNetworkController) reconcileEgressIP(old, new *egressipv1.EgressIP) (err error) {
pkg/ovn/egressip.go:276:func (oc *DefaultNetworkController) reconcileEgressIPNamespace(old, new *v1.Namespace) error {
pkg/ovn/egressip.go:327:func (oc *DefaultNetworkController) reconcileEgressIPPod(old, new *v1.Pod) (err error) {
```

File: go-controller/pkg/ovn/egressip_test.go (hunk 1 above)
Comment (flavio-fernandes) [reply go-controller/pkg/ovn/egressip_test.go:5104]: good idea. ;)

**Comment by flavio-fernandes** [comment]
[FAIL] Kubevirt Virtual Machines when live migration [It] with pre-copy succeeds, should keep connectivity

Flake: https://github.com/ovn-org/ovn-kubernetes/issues/3986

**Review by kyrtapz — APPROVED** [review]
/lgtm

**Review by tssurya — APPROVED** [review]
/lgtm

**Review by dcbw — APPROVED** [review]

**Pull request merged**